### PR TITLE
Add timeout and cancellation check to rescore phase

### LINKED ISSUE #114955: Provide a way for rescore to periodically check for timeouts/cancellation
### Description

Plugin authors and us may be doing a ton of heavy compute work in rescore. However, there is no simple way to check if the original caller is still listening (e.g. did the parent task get cancelled), or if the search has timed out. 

We should provide a way to periodically check for cancellations and timeout. Additionally we should add such checks to our more compute intensive rescoring. 

 - [ ] Add ability to check if the request should be cancelled or timedout
 - [ ] Add periodic checks for the QueryRescorer when it scores documents (this might be slightly difficult and require us to wrap some Lucene related things).
 - [ ] Adjust `LearningToRankRescorer` to periodically check for this during is run

### PR DESCRIPTION
**Pull request by benwtrent**
This adds cancellation checks to rescore phase. This cancellation checks for the parent task being cancelled and for timeout checks.

The assumption is that rescore is always significantly more expensive than a regular query, so we check for timeout as frequently as the most frequent check in ExitableDirectoryReader.

For LTR, we check on hit inference. Maybe we should also check for per feature extraction?

For QueryRescorer, we check in the combine method.

closes: https://github.com/elastic/elasticsearch/issues/114955

### CODE REVIEW

**Comment by benwtrent** [comment]
//cc @ChrisHegarty @dnhatn 

I need to write some tests, but this seems like a good starting point?

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java [hunk 1]
```diff
@@ -979,9 +1001,173 @@ public void testRescoreAfterCollapseRandom() throws Exception {
...
+        int numDocs = indexRandomNumbers("whitespace", -1, false);
+
+        String intToEnglish = English.intToEnglish(between(0, numDocs - 1));
+        String query = intToEnglish.split(" ")[0];
+        assertResponse(
+            prepareSearch().setSearchType(SearchType.QUERY_THEN_FETCH)
+                .setQuery(QueryBuilders.matchQuery("field1", query).operator(Operator.OR))
+                .setSize(10)
+                .addRescorer(new QueryRescorerBuilder(new TimedQueryBuilder(500)).windowSize(100))
+                .setTimeout(TimeValue.timeValueMillis(10)),
```
Comment (dnhatn) [inline server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java:1015]: Using a timeout here can be flaky. Could we use CountDownLatch and task cancellation instead? Also, we can define ScoreFunctionSpec in `SearchPlugin#getScoreFunctions()` to simulate the plugin in question.

**Review by dnhatn — APPROVED** [review]
Production changes look good. Thanks @benwtrent.

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java (hunk 1 above)
Comment (ChrisHegarty) [reply server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java:1015]: I think that we'd prob want both types of tests;
  1) a real timeout that verifies the timeout value is propagated correctly, and 
  2) cancellation of a request that has no explicit timeout (or a large timeout) ?

**Review by ChrisHegarty — APPROVED** [review]

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java (hunk 1 above)
Comment (benwtrent) [reply server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java:1015]: @dnhatn I don't know how to do a countdown latch as this test could serialize. I can add an entirely new suite of tests if we want, but I thought it prudent to get a real integration test in to verify search responses are what we expect (this shows some bugs in my original approach).

**Comment by benwtrent** [comment]
@ChrisHegarty @dnhatn I added a test for verifying we cancel if the task is cancelled in rescore :)

If there isn't any other feed back I am gonna merge & backport this soon.

**Pull request merged**